File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """

        def traverse(current_node):
            if not current_node:
                return None
            
            left_node = traverse(current_node.left)
            right_node = traverse(current_node.right)

            if left_node and right_node:
                return current_node
            
            if current_node == p or current_node == q:
                return current_node

            return left_node if right_node == None else right_node
        
        current_node = traverse(root)
        return current_node
        

File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    TreeNode traverse(TreeNode currNode, TreeNode p, TreeNode q){
        // Base Case:
		if(currNode == null){
			return null;
		}
		
		TreeNode leftNode = traverse(currNode.left, p, q);
		
		TreeNode rightNode = traverse(currNode.right, p, q);
		
		if (currNode == p || currNode == q){
			return currNode;
		}
		else if(leftNode != null && rightNode != null){
			return currNode;
		}
		
		return leftNode == null ? rightNode : leftNode;
	}

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        return traverse(root, p, q);
    }
}